docs/README.md

```diff
@@ -0,0 +1,320 @@
+where you might experience other tools installing the entire `package.json` file.
+
+On the other hand, authoring BUILD files is a chore that's required under Bazel and not in
+other tools like Rush or Nx. We plan a [Gazelle] extension soon which automates much of this toil.
+
+Other recommendations:
+
+-   Put [common flags](https://blog.aspect.dev/bazelrc-flags) in your `.bazelrc` file.
+-   Use [Renovate](https://docs.renovatebot.com/) to keep your Bazel dependencies up-to-date.
+
+### Node.js
+
+rules_js depends on rules_nodejs version 5.0 or greater.
+
+Installation is included in the `WORKSPACE` snippet you pasted from the Installation instructions above.
+
+**API docs:**
+
+-   Installing and choosing the version of Node.js:
+    <https://bazelbuild.github.io/rules_nodejs/install.html>
+-   Rules API: <https://bazelbuild.github.io/rules_nodejs/Core.html>
+-   The Node.js toolchain: <https://bazelbuild.github.io/rules_nodejs/Toolchains.html>
+
+### Use third-party packages from npm
+
+rules_js accesses npm packages using [pnpm].
+pnpm's "virtual store" of packages aligns with Bazel's "external repositories",
+and the pnpm "linker" which creates the `node_modules` tree has semantics we can reproduce with Bazel actions.
+
+If your code works with pnpm, then you should expect it works under Bazel as well.
+This means that if your issue can be reproduced outside of Bazel, using a reproduction with only pnpm,
+then we ask that you fix the issue there, and will close such issues filed on rules_js.
+
+The typical usage is to import an entire `pnpm-lock.yaml` file.
+Create such a file if you don't have one. You could install pnpm on your machine, or use `npx` to run it.
+For example, this command creates a lockfile with minimal installation needed:
+
+```shell
+$ npx pnpm install --lockfile-only
+```
+
+Next, you'll typically use `npm_translate_lock` to translate the Yaml file to Starlark, which Bazel extensions understand.
+The `WORKSPACE` snippet you pasted above already contains this code.
+
+Technically, we run a port of pnpm rather than pnpm itself. The reasoning is as follows:
+
+1. You don't need to install pnpm on your machine to build and test with Bazel.
+1. We re-use pnpm's resolver, by consuming the `pnpm-lock.yaml` file it produces.
+1. We use Bazel's downloader API to fetch package tarballs and extract them to external repositories.
+   To modify the URLs Bazel uses to download packages (for example, to fetch from Artifactory), read
+   <https://blog.aspect.dev/configuring-bazels-downloader>.
+1. We re-use the [`@pnpm/lifecycle`](https://www.npmjs.com/package/@pnpm/lifecycle) package to perform postinstall steps.
+   (These run as cacheable Bazel actions.)
+1. Finally, you link the `node_modules` tree by adding a `npm_link_package` or `npm_link_all_packages` in your `BUILD` file,
+   which populates a tree under `bazel-bin/[path/to/package]/node_modules`.
```

Comment:
I believe this is the most important part of the rulejs core design logic, maybe we can add more information with sequence to describe the flow with more details in the future PR.

Reply:
added a section below about how we use the bazel-bin folder for srcs, genfiles, and deps